### PR TITLE
Add tools to sysroot after build.

### LINKED ISSUE #81431: `cargo fmt` and other tools don't work for a linked toolchain.
This is derived from [this issue](https://github.com/rust-lang/rustup/issues/2548), in which we concluded that the bug should be filed against rustc, as it's an `x.py` problem.

In rustc:
```
$ ./x.py build --stage 1 src/tools/rustfmt/
...
$ find 'build' -name rustfmt
build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/rustfmt
build/x86_64-unknown-linux-gnu/stage1-tools-bin/rustfmt
build/x86_64-unknown-linux-gnu/stage0/share/doc/rustfmt
build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt
build/tmp-rustbuild-tests/main/tmp/tarball/rustfmt
build/tmp-rustbuild-tests/main/tmp/tarball/rustfmt/B/image/share/doc/rustfmt
```

Then in a project using this toolchain via a rustup link:
```
$ cargo fmt
error: 'cargo-fmt' is not installed for the toolchain 'yk-stage1'
note: this is a custom toolchain, which cannot use `rustup component add`
help: if you built this toolchain from source, and used `rustup toolchain link`, then you may be able to build the component with `x.py`                                                                           
```

But note that `cp build/x86_64-unknown-linux-gnu/stage1-tools-bin/rustfmt build/x86_64-unknown-linux-gnu/stage1/bin/rustfmt` does not make this work either, so there is probably more to it?

Thanks

### LINKED ISSUE #97762: `x build src/tools/miri` doesn't install miri as part of the built toolchain
Running `x build src/tools/miri` doesn't include miri as part of the built rust toolchain, meaning that `cargo +custom-toolchain miri` doesn't work

cc @jyn514

### PR DESCRIPTION
**Pull request by WaffleLapkin**
As the title suggest, this makes bootstrap add tools to sysroot after building them. This allows the tools to be used with rustup, e.g.
```shell
; x b rustfmt
; rustfmt +stage1 ./t.rs
```

Resolves #97762
Resolves #81431

-----

r? @jyn514 

I copied the implementation from the same thing with rustdoc ([link](https://github.com/rust-lang/rust/blob/503e19d01e941b88bf6d5b28e9108d046abcfa2d/src/bootstrap/tool.rs#L567-L576)), however there are some questions/problems:

1. [ ] Q: Do we need `miri` in the sysroot? It seems like the only way to run miri is throw `cargo-miri`...
2. [ ] Q: Do we need _both_ `cargo-clippy` and `clippy-driver` in the sysroot?
3. [ ] P: when building any tool, all other tools are deleted for some reason that I can't seem to grasp
     ```shell
     ; x b cargo-miri
     ; ls build/x86_64-unknown-linux-gnu/stage1/bin
     cargo-miri  rustc
     ; x b rustfmt
     ; ls build/x86_64-unknown-linux-gnu/stage1/bin
     rustc  rustfmt
     ; x b clippy
     ; ls build/x86_64-unknown-linux-gnu/stage1/bin
     clippy-driver  rustc
     ```

### CODE REVIEW

**Comment by WaffleLapkin** [comment]
With the power of ~~friendship~~ debug prints, I was able to find out that `Sysroot` step deletes tools from the `bin`.

**Comment by jyn514** [comment]
> when building any tool, all other tools are deleted for some reason

Hmm, I don't recall why we delete the whole sysroot, but copying the whole `stage-N-tools` directly rather than just the newly built binary seems like a reasonable fix :)

**Comment by jyn514** [comment]
> Q: Do we need miri in the sysroot? It seems like the only way to run miri is throw cargo-miri...
> Q: Do we need both cargo-clippy and clippy-driver in the sysroot?

✨ Try it and see ✨ :P can you run `cargo +stage1 clippy` and `cargo +stage1 miri` to see if they work?

I do think we should support miri, but it sounds like you're asking specifically about `miri` vs `cargo-miri`.

**Comment by WaffleLapkin** [comment]
> Hmm, I don't recall why we delete the whole sysroot, but copying the whole `stage-N-tools` directly rather than just the newly built binary seems like a reasonable fix :)

Hm, have you meant `stageN-tools-bin` or `stageN-tools/TARGET/release`? The former doesn't seem to contain everything  (e.g. there is no `cargo-clippy` even though I'm sure I've built it...), while the latter seems to contain too much...

```shell
; ls build/x86_64-unknown-linux-gnu/stage1-tools-bin                             
cargo-miri  clippy-driver  miri  rust-analyzer-proc-macro-srv  rustfmt
; ls build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release
build           cargo-fmt.d    clippy-driver.d  git-rustfmt.d  librustfmt_nightly.d     rust-analyzer-proc-macro-srv    rustfmt-format-diff
cargo-clippy    cargo-miri     deps             incremental    librustfmt_nightly.rlib  rust-analyzer-proc-macro-srv.d  rustfmt-format-diff.d
cargo-clippy.d  cargo-miri.d   examples         libmiri.d      miri                     rustfmt
cargo-fmt       clippy-driver  git-rustfmt      libmiri.rlib   miri.d                   rustfmt.d
```

**Comment by jyn514** [comment]
I meant `stageN-tools-bin`, it's weird that it doesn't have `cargo-clippy`. Maybe dig into why that isn't there? The logic for copying it to stageN-tools-bin is going to be essentially the same as copying it to the sysroot, so I wouldn't spend too much time trying to filter down stuff from `TARGET/release`

**Comment by Mark-Simulacrum** [comment]
I don't think we should copy everything - e.g., compiletest shouldn't get copied into the sysroot. In general, the reason we delete the directory is to make sure that we don't have cases where order of running x.py commands matters.

I would expect the deletion to happen before we copy into the directory though, because it should only happen once per run and we should invoke the relevant Step before copying things into it.

**Comment by WaffleLapkin** [comment]
> I would expect the deletion to happen before we copy into the directory though, [...]

It does happen before the copy, but only the just-built thing and `rustc` are copied. Building `cargo-miri` adds it to the bin, building `rustfmt` after that removes `cago-miri` and adds `rustfmt`, etc.

**Comment by WaffleLapkin** [comment]
> In general, the reason we delete the directory is to make sure that we don't have cases where order of running x.py commands matters.

How does this help? It seems like all files in `stageN/bin` are independent from each other, so even if we don't delete the directory, it'll be still fine

**Comment by Mark-Simulacrum** [comment]
> How does this help? It seems like all files in stageN/bin are independent from each other, so even if we don't delete the directory, it'll be still fine

I don't think this is necessarily true (e.g., you might have one binary executing another), and it's much easier for something that is kept around persistently to e.g. have a stale binary that you don't realize should have been rebuilt and giving you very confusing results.

> It does happen before the copy, but only the just-built thing and rustc are copied. Building cargo-miri adds it to the bin, building rustfmt after that removes cago-miri and adds rustfmt, etc.

Presumably this doesn't happen if you build both at once, e.g., `x.py build src/tools/miri src/tools/rustfmt`? It doesn't seem unreasonable to me to expect the user to do that, and not try to guess that you "still want" the old miri (particularly if that means e.g. that an edit to rustc either leaves a stale miri in place or magically detects it and starts building miri even though you've not asked for it).

**Comment by WaffleLapkin** [comment]
`./x.py build src/tools/miri src/tools/rustfmt` does work, but I don't think it's reasonable to expect users to do that. I personally would build tools when I need them, and if I first need tool-A, then tool-B, then I would run `x b a`, then use tool-A, then `x b b` and then I would be confused why tool-A stopped working.

IMO it would make more sense if we would clear the directory only when `rustc` is **rebuilt**.
```shell
; x b compiler # clears the directory
; x b miri     # leaves the directory as-is
; x b rustfmt  # leaves the directory as-is

; nano compiler/... # change compiler
; x b miri          # compiler is rebuilt => clears the directory
```

**Comment by Mark-Simulacrum** [comment]
I suppose that's possible, but I'm typically very wary of conditional clearing for that kind of directory. It's currently the case that all sysroots managed by rustbuild are always re-created on every build (AFAIK), which is a really convenient property for the implementation. In most cases if the goal is a "stable" sysroot, I sort of want to say "x.py install" or similar is the right way to get that, not trying to make the existing sysroots more durable.

Part of the reason I worry here is that "only when rustc is rebuilt" is a pretty annoying property to actually get; we don't directly know in rustbuild when something is rebuilt, cargo does caching for us. We do have some clear_if_dirty and such, with stamps, but that's really something that we're trying to move away from over time because getting them right is really tricky. (For an example: changing codegen can mean that you've not modified the rustc binary, so its timestamp hasn't changed, but the librustc_codegen_llvm.so has, and we're not tracking that, so you end up with an inconsistent state where newly built binaries (e.g., tools, etc.) are linking against a potentially old std built by the old rustc. It is much simpler to always re-hard-link everything and gives you a lot of nice properties as a result.

**Comment by jyn514** [comment]
@WaffleLapkin let's require for now that people need to use `build rustfmt miri` to get both at once; if we agree on a way to make that simpler in the future we can make it better then. Having *a* way to get the tools is better than nothing :)

File: src/bootstrap/tool.rs [hunk 1]
```diff
@@ -853,7 +854,18 @@ macro_rules! tool_extended {
...
-                })
+                })?;
+
+                if (false $(|| $add_to_sysroot)?) && $sel.compiler.stage > 0 {
+                    let bin_dir = $builder.sysroot($sel.compiler).join("bin");
+                    t!(fs::create_dir_all(&bin_dir));
+                    let bin_tool = bin_dir.join(exe($tool_name, $sel.compiler.host));
+                    let _ = fs::remove_file(&bin_tool);
+                    $builder.copy(&tool, &bin_tool);
+                    Some(bin_tool)
```
Comment (jyn514) [inline src/bootstrap/tool.rs:865]: hmm, why is the path we return from the Step different depending on whether it's in the sysroot or not? I would expect this to be consistent with ToolBuild even if we copy the tool to the sysroot as a side effect.

**Review by jyn514 — APPROVED** [review]
Thanks! LGTM other than the one question :)

File: src/bootstrap/tool.rs (hunk 1 above)
Comment (WaffleLapkin) [reply src/bootstrap/tool.rs:865]: We do this for rustdoc for some reason: https://github.com/rust-lang/rust/blob/503e19d01e941b88bf6d5b28e9108d046abcfa2d/src/bootstrap/tool.rs#L567-L576

I copied this from there

Comment (jyn514) [reply src/bootstrap/tool.rs:865]: Fascinating. ok, seems good to be consistent.

**Comment by WaffleLapkin** [comment]
I can't get `miri` working, after `x build miri cargo-miri`, `cargo +stage1 miri test` fails with the following:
```rust
thread 'main' panicked at 'failed to determine underlying rustc version of Miri: CommandError { stdout: "", stderr: "/home/waffle/rust-b/build/x86_64-unknown-linux-gnu/stage1/bin/miri: error while loading shared libraries: librustc_driver-152a9025dac882e1.so: cannot open shared object file: No such file or directory\n" }', src/tools/miri/cargo-miri/src/util.rs:116:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by WaffleLapkin** [comment]
clippy doesn't work either, although for a different reason. `x build clippy` only builds (?copied) `clippy-driver`, but not `cargo-clippy`, while `x build cargo-clippy` doesn't work at all: 
```rust
error: no `build` rules matched ["cargo-clippy"]
help: run `x.py build --help --verbose` to show a list of available paths
note: if you are adding a new Step to bootstrap itself, make sure you register it with `describe!`
Build completed unsuccessfully in 0:00:00
```

**Comment by WaffleLapkin** [comment]
`clippy` mystery solved:  both `cargo-clippy` and `clippy-driver` use the same path: https://github.com/rust-lang/rust/blob/91931ec2fcb67a8e61080a97345c828a875c86ec/src/bootstrap/tool.rs#L869-L870

We then only allow one thing to be called with a given path: https://github.com/rust-lang/rust/blob/91931ec2fcb67a8e61080a97345c828a875c86ec/src/bootstrap/builder.rs#L316-L326

So only the first one (happens to be `clippy-driver`) is being built.

If I change paths to be different bootstrap tries to build both (I say tries, because `error: manifest path `/home/waffle/rust-b/src/tools/clippy/src/driver.rs/Cargo.toml` does not exist`).

Not sure what to do with this though

**Comment by Mark-Simulacrum** [comment]
r=me, seems okay. I don't think there's anything inherently wrong in multiple things sharing the same path, though it will further hinder the efforts to make --exclude and such actually run (or stop running) a specific thing.

**Comment by jyn514** [comment]
Great, thank you :)

@WaffleLapkin this is ready to merge once you fix the conflicts :)

**Comment by jyn514** [comment]
@WaffleLapkin I found that even clippy has the sysroot bug you mentioned for miri - I think we need to be doing some extra magic (maybe with `-rpath`?) to make sure clippy can find the sysroot.

@bors r- this isn't useful as-is

**Comment by jyn514** [comment]
cc https://rust-lang.zulipchat.com/#narrow/stream/269128-miri/topic/building.20miri.20out.20of.20a.20rustc.20branch/near/317849329, https://github.com/rust-lang/rust/issues/53650, https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/Stage.20numbering.20for.20tools

**Pull request closed without merging**